python/tvm/octo/octo_model.py

```diff
@@ -0,0 +1,191 @@
+# under the License.
+# pylint: disable=invalid-name, wrong-import-position
+"""Wrapper class for compiled models."""
+import json
+import tarfile
+import numpy as np
+from pathlib import Path
+from typing import Optional, Union, Dict, Tuple, List
+import tvm
+from tvm import relax
+from tvm.contrib import utils
+
+
+class OctoModel(object):
+    """A compiled model wrapper that provides helpful utilities.
+
+    Parameters
+    ----------
+    exe : Optional[relax.vm.Executable]
+        A compiled executable that can be loaded and run by a relax VM.
+    input_info : Optional[Dict[str, Tuple[List, str]]]
+        Information about the input names, shapes, and types for the VM.
+        Will be loaded from memory if possible.
+    model_path : Optional[Union[str, Path]]
+        The path to a saved OctoModel, one of exe and model_path must
+        be specified.
+    target : Optional[tvm.target.Target]
+        The target being compiled for.
+    """
+
+    def __init__(
+        self,
+        exe: Optional[relax.vm.Executable] = None,
+        input_info: Optional[Dict[str, Tuple[List, str]]] = None,
+        model_path: Optional[Union[str, Path]] = None,
+        target: Optional[tvm.target.Target] = None,
+    ):
+        self.target = target
+
+        if exe is None and model_path is None:
+            raise ValueError("One of vm and model_path must be provided.")
+
+        self._tmp_dir = utils.tempdir()
+
+        if model_path is not None:
+            exe, input_info = self.load(model_path)
+
+        self.dev = tvm.device(self.target.get_target_device_type())
+        self.exe = exe
+        self.input_info = input_info
+
+        # Create a vm from exe.
+        self.vm = relax.VirtualMachine(self.exe, self.dev, profile=True)
+
+    def save(
+        self, model_path: Union[str, Path]
+    ) -> Tuple[relax.vm.Executable, Dict[str, relax.StructInfo]]:
+        """Save the OctoModel to disk.
+
+        Parameters
+        ----------
+        model_path : Union[str, Path]
+            A full path to save this OctoModel to including the output file name.
+            The file will be saved as a tar file so using a ".tar" extension is advised.
+        """
+        # Only two artifacts need to be saved, the exe and the input struct info.
+        # Serialize both to a temp directory.
+        exe_path = self._tmp_dir.relpath("exe.so")
+        self.exe.mod.export_library(exe_path)
+        input_info_path = self._tmp_dir.relpath("input_info.json")
+        with open(input_info_path, "w") as fo:
+            json.dump(self.input_info, fo)
+
+        # Save additional metadata.
+        metadata = {"target": str(self.target)}
+        metadata_path = self._tmp_dir.relpath("metadata.json")
+        with open(metadata_path, "w") as fo:
+            json.dump(metadata, fo)
+
+        # Tar the tempfile and save to the designated model_path.
+        with tarfile.open(model_path, "w") as tar:
```

Comment:
should we recycle Model Library Format here to avoid creating two of these?

Reply:
I think using MLF is a good idea but not currently applicable since it is tied to `micro` and would need to be extended to support Relax. I'd be in favor of bumping MLF to a tvm namescoped standard that can be used more broadly. For now, I'll just note the format more clearly in the comments and we can change later.